src/main/java/edu/bu/cs633/minimalsearchengine/crawler/MSEWebCrawler.java

```diff
@@ -0,0 +1,175 @@
+package edu.bu.cs633.minimalsearchengine.crawler;
+
+import edu.bu.cs633.minimalsearchengine.models.dao.Page;
+import edu.bu.cs633.minimalsearchengine.models.dao.Word;
+import edu.bu.cs633.minimalsearchengine.utils.Utilities;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+public class MSEWebCrawler {
+
+    private String url;
+
+    public static final int MAX_LINKS = 3;
+    private static final String DELIMITERS = "[\\s\\W]+";
+
+    private static final Logger logger = Logger.getLogger(MSEWebCrawler.class.toString());
+
+    public MSEWebCrawler() {
+    }
+
+    public MSEWebCrawler(String url) {
+        this.url = url;
+    }
+
+    public String getUrl() {
+        return url;
+    }
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
+
+    public Set<Page> crawl() throws IOException {
+
+        Set<Page> pages = findLinks();
+
+        if (pages == null || pages.size() == 0) {
+            throw new IOException("No pages or invalid URL detected.");
+        }
+
+        return extractPagesAndWords(pages);
+    }
+
+    /**
+     * Finds 2 additional URLs from the requested URL which needs to be crawled.
+     * @return A set of Pages with URLs which needs to be crawled.
+     * @throws IOException occurs when tries to read the URL
+     */
+    private Set<Page> findLinks() throws IOException {
+
+        boolean isValidURL;
+
+        Set<Page> pages = new HashSet<>();
+
+        try {
+            isValidURL = Utilities.isValidURL(getUrl());
```

Comment:
@HamidurRahman1 
Not really an issue but I'm curious about why you're calling `getUrl()` here rather than just accessing the `url` directly from this private method. Is this considered best practice?

Reply:
Utilize instance methods wherever possible to mitigate risk in many ways. One of the common issue is find-and-replace. Also, internally you can name the variable mySecretURLToBeCrawled but still have a standard method name to outsiders